GitUI/Avatars/AvatarService.cs

```diff
@@ -1,12 +1,132 @@
+
+            HotSwapProvider = new HotSwapAvatarProvider();
+            UpdateAvatarProvider();
+
+            Default = WrapInCacheAndStaticFallback(HotSwapProvider);
+        }
+
+        public static ICachedAvatarProvider Default { get; }
+
+        /// <summary>
+        /// Updates the internal avatar provider chain to
+        /// reflect the current active (according to <see cref="AppSettings"/> provider.
+        /// </summary>
+        public static void UpdateAvatarProvider()
+        {
+            HotSwapProvider.Provider = CreateAvatarProvider(AppSettings.AvatarProvider, AppSettings.AvatarFallbackType);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="IAvatarProvider"/> based on the given
+        /// <see cref="AvatarProvider"/> and <see cref="AvatarFallbackType"/> enum options.
+        /// </summary>
+        /// <returns>Returns an avatar provider that can be used to resolve user avatars.</returns>
+        public static IAvatarProvider CreateAvatarProvider(
+            AvatarProvider provider,
+            AvatarFallbackType? fallbackType,
+            IAvatarDownloader downloader = null)
+        {
+            // initialize download only if needed (some options, like local providers, don't need a downloader)
+            // and use the downloader provided as parameter if possible.
+            var lazyDownloader = new Lazy<IAvatarDownloader>(() => downloader ?? new AvatarDownloader());
+
+            // build collection of (non-null) providers
+            var providers = new[]
+            {
+                BuildMainProvider(),
+                BuildFallbackProvider()
+            }
+            .Where(p => p != null)
+            .ToArray();
+
+            // only create chained avatar overhead if really needed
+            return providers.Length switch
+            {
+                0 => null,
+                1 => providers[0],
+                _ => new ChainedAvatarProvider(providers),
+            };
+
+            // Local methods to build requested main and fallback providers:
+
+            IAvatarProvider BuildMainProvider()
+            {
+                return provider switch
+                {
+                    AvatarProvider.Default => BuildDefaultMainProvider(),
+                    AvatarProvider.Custom => CustomAvatarProvider.ParseTemplateString(AppSettings.CustomAvatarTemplate, lazyDownloader.Value),
+                    _ => null,
+                };
+            }
+
+            IAvatarProvider BuildDefaultMainProvider()
+            {
+                // 1. query github (with non-reply and regular email addresses)
+                //    github might internally fall back to gravatar, so only need a single request in most cases.
+
+                // 2. resolve via gravatar (for users that don't have a github account)
+                //    this request also directly provides the fallback if it's gravatar compatible.
```

Comment:
[nit] Need to correct casing, e.g. GitHub, Gravatar.

Reply:
✔ done. May I ask what _"[nit]"_ stands for?

Reply:
nit pick: https://en.wiktionary.org/wiki/nitpick

Reply:
oooh 🤦‍♂️, I googled it and just found a bunch strange abbreviations. Thank you

Reply:
Thank you @gerhardol. 

@wischi-chr _nitpicks_ or _nits_ for short are usually things that would be good to fix/correct but generally of non-blocking nature.